source/PlayerInfo.cpp

```diff
@@ -1849,6 +1862,14 @@ const map<string, int64_t> &PlayerInfo::Conditions() const
 
 
 
+// Uuid for the gifted ships, with the ship class follow by the names they had when they were gifted to the player.
+const std::map<std::string, EsUuid> &PlayerInfo::GiftedShips() const
+{
+	return giftedShips;
+}
```

Comment:
To be honest I don't understand why we have such a complicated system for so little practical applications (this is the first one for them as far as I could find, and a simple number would have worked just fine), system which I do not fully understand, and I did not try to make tests before so this may take a while. 
I will try to do this, just keeping it for last.

Reply:
> I don't think I made EsUuid persistent when copied, so putting it into a container by value is likely unsafe.
> 
> I expect you to add unit tests proving that one can iterate over a collection with this signature and use it as you propose to do. Identifying iteration patterns that are unsafe could also be useful.
> 
> (The copy semantics were chosen to limit the need for all contributors to understand when the game was copying ships with the intent of using the copy as a new item and remembering to null the copy's ID, as that is the intent in almost every single instance.)

I'm really bad at this, I simply never did a unit test before.
Lines 172-175 of sts/src/test_esuuid.cpp mention that a ship gifted at one point should be identifiable later, but if not with Uuid, I cannot see how. So with that in mind it would be reasonable to assume it is already tested for in the following unit tests?
If I understand correctly, the Uuid is never changed except when a new ship is bought from a model? (But maybe this practical justification is bad because in theory it could go wrong?)
In any cases, I am not sure how could this iteration be a problem, as in what specifically should I test?

Reply:
Actually, looking at this more, ESUUID is simply always constant for ships that are owned by the player, since its written in the savefile. Changing them in game to write them differently again would simply be weird, and I don't expect that to ever happen.